### PR TITLE
feat: Mongo transaction

### LINKED ISSUE #200: [Task] Mongo transaction
## What should be done 
We want the dev to be able to start a new transaction and run mongo query with it using api request.

https://github.com/zookzook/elixir-mongodb-driver

### PR DESCRIPTION
**Pull request by pichoemr**
#200 

## Checklist
- [ ] I didn't over-scope my PR
- [ ] My PR title matches the [commit convention](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] I did not include breaking changes
- [ ] I included unit tests that cover my changes
- [ ] I added/updated the documentation about my changes
- [ ] I made my own code-review before requesting one

### CODE REVIEW

File: config/test.exs [hunk 1]
```diff
@@ -6,7 +6,8 @@ config :application_runner,
   ecto_repos: [ApplicationRunner.Repo],
   faas_url: "http://localhost:1234",
   faas_auth: "Basic YWRtaW46M2kwREc4NTdLWlVaODQ3R0pheW5qMXAwbQ==",
-  faas_registry: "registry.gitlab.com/lenra/platform/lenra-ci"
+  faas_registry: "registry.gitlab.com/lenra/platform/lenra-ci",
+  env: "test"
```
Comment (jonas-martinez) [inline config/test.exs:10]: Why did you add this ?

File: lib/errors/business_error.ex [hunk 2]
```diff
@@ -19,6 +19,9 @@ defmodule ApplicationRunner.Errors.BusinessError do
       {:incorrect_view_mode, "The view mode should be one of 'lenra', 'json'."},
       {:no_action_in_listener, "Your listener does not have the required property 'action'"},
       {:route_does_not_exist, "The given route does not exist. Please check your manifest."},
-      {:invalid_channel_name, "The given channel name does not exist."}
+      {:invalid_channel_name, "The given channel name does not exist."},
+      {:error_during_transaction_start,
+       "An error occurs during transaction start, please try again or contact us"},
```
Comment (jonas-martinez) [inline lib/errors/business_error.ex]:
```suggestion
       "An error occured during transaction start."},
```

**Review by jonas-martinez — CHANGES_REQUESTED** [review]

File: lib/mongo_storage.ex [hunk 3]
```diff
@@ -174,6 +198,16 @@ defmodule ApplicationRunner.MongoStorage do
     end
   end
 
+  defp handle_delete_doc(pipe, session_uuid, coll, bson_doc_id) do
+    if session_uuid != nil do
+      pipe
+      |> Mongo.delete_one(coll, %{"_id" => bson_doc_id}, session: Swarm.whereis_name(session_uuid))
```
Comment (jonas-martinez) [inline lib/mongo_storage.ex:216]: Do you really need to pass the session uuid in the delete_one method ? I don't think this is necessary because you already give the _id which should be enough to reference the doc. Maybe I am missing something, what do you think?

**Review by jonas-martinez — CHANGES_REQUESTED** [review]

File: lib/mongo_storage.ex [hunk 4]
```diff
@@ -228,4 +262,50 @@ defmodule ApplicationRunner.MongoStorage do
...
+  end
+
+  def commit_transaction(session_uuid, env_id) do
+    with :ok <- Mongo.Session.commit_transaction(Swarm.whereis_name(session_uuid)),
+         :ok <-
+           Mongo.Session.end_session(mongo_instance(env_id), Swarm.whereis_name(session_uuid)),
+         :ok <- Swarm.unregister_name(session_uuid) do
+      :ok
+    else
+      # TODO: check what to do in this case
```
Comment (jonas-martinez) [inline lib/mongo_storage.ex:310]: It will depend on the error that is triggered.

If the commit_transaction failed, we can let the user run his transaction again.
If the end_session failed, we should find why and log it, maybe the session crashed after the commit_transaction and there is no need to do anything. In theory, if the commit_transaction is successful, the data should be properly inserted.
If the unregister_name failed, what do we want to do ?

File: lib/mongo_storage.ex [hunk 5]
```diff
@@ -228,4 +262,50 @@ defmodule ApplicationRunner.MongoStorage do
...
+    end
+  end
+
+  def revert_transaction(session_uuid, env_id) do
+    with :ok <-
+           Mongo.Session.end_session(mongo_instance(env_id), Swarm.whereis_name(session_uuid)),
+         :ok <- Swarm.unregister_name(session_uuid) do
+      :ok
+    else
+      # TODO: check what to do in this case
```
Comment (jonas-martinez) [inline lib/mongo_storage.ex:322]: I'm not sure about this one. 
Does the end_session make sure that the data is not inserted into the db ? 
If that is the case we should handle the else like this:

if end_session failed, find why and log it, maybe the session crashed and in this case it is reverted by default. But if the session is still open, we must make sure that it is deleted.
If the unregister_name failed, what do we want to do ?

File: test/mongo_storage_test.exs [hunk 6]
```diff
@@ -0,0 +1,19 @@
+defmodule ApplicationRunner.MongoStorageTest do
+  use ApplicationRunner.ConnCase, async: false
+
+  alias ApplicationRunner.MongoStorage
+  alias ApplicationRunner.Errors.BusinessError
+  alias ApplicationRunner.Environment.MongoInstance
+
+  defp setup_mongo(env_id) do
```
Comment (jonas-martinez) [inline test/mongo_storage_test.exs]: More tests needs to be done here to make sure that this feature is 100% working as it can cause a lot of problems in the application data.

**Review by jonas-martinez — CHANGES_REQUESTED** [review]

File: lib/mongo_storage.ex (hunk 3 above)
Comment (taorepoara) [reply lib/mongo_storage.ex:216]: It's the transaction uuid which is needed to be hable to rollback the all transaction commands.

File: config/test.exs (hunk 1 above)
Comment (pichoemr) [reply config/test.exs:10]: This is for mongo databse name

File: test/mongo_storage_test.exs (hunk 6 above)
Comment (pichoemr) [reply test/mongo_storage_test.exs]: Of course, I had some problems doing the tests so I left them on the side for now, I'll do them as soon as possible!

**Review by jonas-martinez — APPROVED** [review]
